# Data Science Unit 1 Sprint Challenge 4

## Exploring Data, Testing Hypotheses

In this sprint challenge you will look at a dataset of people being approved or rejected for credit.

https://archive.ics.uci.edu/ml/datasets/Credit+Approval

Data Set Information: This file concerns credit card applications. All attribute names and values have been changed to meaningless symbols to protect confidentiality of the data. This dataset is interesting because there is a good mix of attributes -- continuous, nominal with small numbers of values, and nominal with larger numbers of values. There are also a few missing values.

Attribute Information:
- A1: b, a.
- A2: continuous.
- A3: continuous.
- A4: u, y, l, t.
- A5: g, p, gg.
- A6: c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff.
- A7: v, h, bb, j, n, z, dd, ff, o.
- A8: continuous.
- A9: t, f.
- A10: t, f.
- A11: continuous.
- A12: t, f.
- A13: g, p, s.
- A14: continuous.
- A15: continuous.
- A16: +,- (class attribute)

Yes, most of that doesn't mean anything. A16 is a variable that indicates whether or not a person's request for credit has been approved or denied. This is a good candidate for a y variable since we might want to use the other features to predict this one. The remaining variables have been obfuscated for privacy - a challenge you may have to deal with in your data science career.

Sprint challenges are evaluated based on satisfactory completion of each part. It is suggested you work through it in order, getting each aspect reasonably working, before trying to deeply explore, iterate, or refine any given step. Once you get to the end, if you want to go back and improve things, go for it!

## Part 1 - Load and validate the data

- Load the data as a `pandas` data frame.
- Validate that it has the appropriate number of observations (you can check the raw file, and also read the dataset description from UCI).
- UCI says there should be missing data - check, and if necessary change the data so pandas recognizes it as na
- Make sure that the loaded features are of the types described above (continuous values should be treated as float), and correct as necessary

This is review, but skills that you'll use at the start of any data exploration. Further, you may have to do some investigation to figure out which file to load from - that is part of the puzzle.

Hint: If a column has the datatype of "object" even though it's made up of float or integer values, you can coerce it to act as a numeric column by using the `pd.to_numeric()` function.

In [0]:
#Load the data as a pandas data frame
import pandas as pd
import numpy as np
from scipy import stats
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
url='https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data'

df = pd.read_csv(url,header=None,
                 names=['A1','A2','A3','A4','A5','A6','A7',
                        'A8','A9','A10','A11','A12','A13','A14','A15','A16'])
df.head()

A1     A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  00202    0   +
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g  00043  560   +
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  00280  824   +
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  00100    3   +
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  00120    0   +

In [0]:
#Validate that it has the appropriate number of observations (you can check the raw file, and also read the dataset description from UCI).
df.shape

(690, 16)

In [0]:
#UCI says there should be missing data - check, and if necessary change 
#the data so pandas recognizes it as na
df.isna().sum()

A1     0
A2     0
A3     0
A4     0
A5     0
A6     0
A7     0
A8     0
A9     0
A10    0
A11    0
A12    0
A13    0
A14    0
A15    0
A16    0
dtype: int64

In [0]:
df.nunique()

A1       3
A2     350
A3     215
A4       4
A5       4
A6      15
A7      10
A8     132
A9       2
A10      2
A11     23
A12      2
A13      3
A14    171
A15    240
A16      2
dtype: int64

In [0]:
# A1:  12
#     A2:  12
#     A4:   6
#     A5:   6
#     A6:   9
#     A7:   9
#     A14: 13

In [0]:
df['A1'].value_counts()

b    468
a    210
?     12
Name: A1, dtype: int64

In [0]:
df = df.replace({'?':np.NaN})
df.head()

A1     A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  00202    0   +
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g  00043  560   +
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  00280  824   +
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  00100    3   +
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  00120    0   +

In [0]:
df.isna().sum()

A1     12
A2     12
A3      0
A4      6
A5      6
A6      9
A7      9
A8      0
A9      0
A10     0
A11     0
A12     0
A13     0
A14    13
A15     0
A16     0
dtype: int64

In [0]:
# Make sure that the loaded features are of the types described above (continuous values should be treated as float), 
# and correct as necessary
df.dtypes
    # A1:	b, a.
    # A2:	continuous. 
    # A3:	continuous.
    # A4:	u, y, l, t.
    # A5:	g, p, gg.
    # A6:	c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff.
    # A7:	v, h, bb, j, n, z, dd, ff, o.
    # A8:	continuous.
    # A9:	t, f.
    # A10:	t, f.
    # A11:	continuous.
    # A12:	t, f.
    # A13:	g, p, s.
    # A14:	continuous.
    # A15:	continuous.
    # A16: +,-         (class attribute)


A1      object
A2      object
A3     float64
A4      object
A5      object
A6      object
A7      object
A8     float64
A9      object
A10     object
A11      int64
A12     object
A13     object
A14     object
A15      int64
A16     object
dtype: object

In [0]:
    # A2:	continuous.   #Change 
    # A3:	continuous.   #OK
    # A8:	continuous.   #OK
    # A11:	continuous. #Change
    # A14:	continuous. #Change 
    # A15:	continuous. #Change
df['A2']=df['A2'].astype(float)
df['A8']=df['A8'].astype(float)
df['A11']=df['A11'].astype(float)
df['A15']=df['A15'].astype(float)

df.dtypes

A1      object
A2     float64
A3     float64
A4      object
A5      object
A6      object
A7      object
A8     float64
A9      object
A10     object
A11    float64
A12     object
A13     object
A14     object
A15    float64
A16     object
dtype: object

## Part 2 - Exploring data, Testing hypotheses

The only thing we really know about this data is that A16 is the class label. Besides that, we have 6 continuous (float) features and 9 categorical features.

Explore the data: you can use whatever approach (tables, utility functions, visualizations) to get an impression of the distributions and relationships of the variables. In general, your goal is to understand how the features are different when grouped by the two class labels (`+` and `-`).

For the 6 continuous features, how are they different when split between the two class labels? Choose two features to run t-tests (again split by class label) - specifically, select one feature that is *extremely* different between the classes, and another feature that is notably less different (though perhaps still "statistically significantly" different). You may have to explore more than two features to do this.

For the categorical features, explore by creating "cross tabs" (aka [contingency tables](https://en.wikipedia.org/wiki/Contingency_table)) between them and the class label, and apply the Chi-squared test to them. [pandas.crosstab](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.crosstab.html) can create contingency tables, and [scipy.stats.chi2_contingency](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) can calculate the Chi-squared statistic for them.

There are 9 categorical features - as with the t-test, try to find one where the Chi-squared test returns an extreme result (rejecting the null that the data are independent), and one where it is less extreme.

**NOTE** - "less extreme" just means smaller test statistic/larger p-value. Even the least extreme differences may be strongly statistically significant.

Your *main* goal is the hypothesis tests, so don't spend too much time on the exploration/visualization piece. That is just a means to an end - use simple visualizations, such as boxplots or a scatter matrix (both built in to pandas), to get a feel for the overall distribution of the variables.

This is challenging, so manage your time and aim for a baseline of at least running two t-tests and two Chi-squared tests before polishing. And don't forget to answer the questions in part 3, even if your results in this part aren't what you want them to be.

In [0]:
df['A16'] = df['A16'].replace({'+':'positive', '-':'negative'})
df.head()

A1     A2     A3 A4 A5 A6 A7  ...  A10  A11 A12  A13    A14    A15       A16
0  b  30.83  0.000  u  g  w  v  ...    t  1.0   f    g  00202    0.0  positive
1  a  58.67  4.460  u  g  q  h  ...    t  6.0   f    g  00043  560.0  positive
2  a  24.50  0.500  u  g  q  h  ...    f  0.0   f    g  00280  824.0  positive
3  b  27.83  1.540  u  g  w  v  ...    t  5.0   t    g  00100    3.0  positive
4  b  20.17  5.625  u  g  w  v  ...    f  0.0   f    s  00120    0.0  positive

[5 rows x 16 columns]

In [0]:
positive= df[df['A16'] == "positive"]#split data class 
print(positive.shape)
positive.head()

(307, 16)


A1     A2     A3 A4 A5 A6 A7  ...  A10  A11 A12  A13    A14    A15       A16
0  b  30.83  0.000  u  g  w  v  ...    t  1.0   f    g  00202    0.0  positive
1  a  58.67  4.460  u  g  q  h  ...    t  6.0   f    g  00043  560.0  positive
2  a  24.50  0.500  u  g  q  h  ...    f  0.0   f    g  00280  824.0  positive
3  b  27.83  1.540  u  g  w  v  ...    t  5.0   t    g  00100    3.0  positive
4  b  20.17  5.625  u  g  w  v  ...    f  0.0   f    s  00120    0.0  positive

[5 rows x 16 columns]

In [0]:
negative= df[df['A16'] == "negative"]#split data class 
print(negative.shape)
negative.head()

(383, 16)


A1     A2   A3 A4 A5  A6  A7  ...  A10  A11 A12  A13    A14  A15       A16
70  b  32.33  7.5  u  g   e  bb  ...    f  0.0   t    s  00420  0.0  negative
71  b  34.83  4.0  u  g   d  bb  ...    f  0.0   t    g    NaN  0.0  negative
72  a  38.58  5.0  u  g  cc   v  ...    f  0.0   t    g  00980  0.0  negative
73  b  44.25  0.5  u  g   m   v  ...    f  0.0   f    s  00400  0.0  negative
74  b  44.83  7.0  y  p   c   v  ...    f  0.0   f    g  00160  2.0  negative

[5 rows x 16 columns]

In [0]:
pd.crosstab([df['A16']],df['A1'] ,margins=True )

A1          a    b  All
A16                    
negative  112  262  374
positive   98  206  304
All       210  468  678

In [0]:
df.describe()

A2          A3          A8        A11            A15
count  678.000000  690.000000  690.000000  690.00000     690.000000
mean    31.568171    4.758725    2.223406    2.40000    1017.385507
std     11.957862    4.978163    3.346513    4.86294    5210.102598
min     13.750000    0.000000    0.000000    0.00000       0.000000
25%     22.602500    1.000000    0.165000    0.00000       0.000000
50%     28.460000    2.750000    1.000000    0.00000       5.000000
75%     38.230000    7.207500    2.625000    3.00000     395.500000
max     80.250000   28.000000   28.500000   67.00000  100000.000000

In [0]:
df.describe(exclude='number')

A1   A4   A5   A6   A7   A9  A10  A12  A13    A14       A16
count   678  684  684  681  681  690  690  690  690    677       690
unique    2    3    3   14    9    2    2    2    3    170         2
top       b    u    g    c    v    t    f    f    g  00000  negative
freq    468  519  519  137  399  361  395  374  625    132       383

In [0]:
# A16 == Class
#A1= Sex (a: male 1 b: female 0)

df['A1'] = df['A1'].replace({'a':'1', 'b':'0'})
df.head()


A1     A2     A3 A4 A5 A6 A7  ...  A10  A11 A12  A13    A14    A15       A16
0  0  30.83  0.000  u  g  w  v  ...    t  1.0   f    g  00202    0.0  positive
1  1  58.67  4.460  u  g  q  h  ...    t  6.0   f    g  00043  560.0  positive
2  1  24.50  0.500  u  g  q  h  ...    f  0.0   f    g  00280  824.0  positive
3  0  27.83  1.540  u  g  w  v  ...    t  5.0   t    g  00100    3.0  positive
4  0  20.17  5.625  u  g  w  v  ...    f  0.0   f    s  00120    0.0  positive

[5 rows x 16 columns]

In [0]:
#df['A1']=df['A1'].astype(float)

In [0]:
# df['A16'] = df['A16'].replace({'positive':'1', 'negative':'0'})

In [0]:
positive= df[df['A16'] == 'positive']#split data class 
print(positive.shape)
positive.head()

(307, 16)


A1     A2     A3 A4 A5 A6 A7  ...  A10  A11 A12  A13    A14    A15       A16
0  b  30.83  0.000  u  g  w  v  ...    t  1.0   f    g  00202    0.0  positive
1  a  58.67  4.460  u  g  q  h  ...    t  6.0   f    g  00043  560.0  positive
2  a  24.50  0.500  u  g  q  h  ...    f  0.0   f    g  00280  824.0  positive
3  b  27.83  1.540  u  g  w  v  ...    t  5.0   t    g  00100    3.0  positive
4  b  20.17  5.625  u  g  w  v  ...    f  0.0   f    s  00120    0.0  positive

[5 rows x 16 columns]

In [0]:
negative= df[df['A16'] == 'negative']#split data class 
print(negative.shape)
negative.head()

(383, 16)


A1     A2   A3 A4 A5  A6  A7      A8 A9 A10  A11 A12 A13    A14  A15  A16
70  0.0  32.33  7.5  u  g   e  bb   1.585  t   f  0.0   t   s  00420  0.0  0.0
71  0.0  34.83  4.0  u  g   d  bb  12.500  t   f  0.0   t   g    NaN  0.0  0.0
72  1.0  38.58  5.0  u  g  cc   v  13.500  t   f  0.0   t   g  00980  0.0  0.0
73  0.0  44.25  0.5  u  g   m   v  10.750  t   f  0.0   f   s  00400  0.0  0.0
74  0.0  44.83  7.0  y  p   c   v   1.625  f   f  0.0   f   g  00160  2.0  0.0

In [0]:
man= df[df['A1'] == 'a']#split data sex
print(man.shape)
man.head()

(210, 16)


A1     A2      A3 A4 A5  A6 A7  ...  A10  A11 A12  A13    A14     A15       A16
1   a  58.67   4.460  u  g   q  h  ...    t  6.0   f    g  00043   560.0  positive
2   a  24.50   0.500  u  g   q  h  ...    f  0.0   f    g  00280   824.0  positive
7   a  22.92  11.585  u  g  cc  v  ...    f  0.0   f    g  00080  1349.0  positive
12  a  38.25   6.000  u  g   k  v  ...    f  0.0   t    g  00000     0.0  positive
14  a  45.83  10.500  u  g   q  v  ...    t  7.0   t    g  00000     0.0  positive

[5 rows x 16 columns]

In [0]:
woman= df[df['A1'] == 'b']#split data sex
print(woman.shape)
woman.head()

(468, 16)


A1     A2     A3 A4 A5 A6 A7  ...  A10  A11 A12  A13    A14      A15       A16
0  b  30.83  0.000  u  g  w  v  ...    t  1.0   f    g  00202      0.0  positive
3  b  27.83  1.540  u  g  w  v  ...    t  5.0   t    g  00100      3.0  positive
4  b  20.17  5.625  u  g  w  v  ...    f  0.0   f    s  00120      0.0  positive
5  b  32.08  4.000  u  g  m  v  ...    f  0.0   t    g  00360      0.0  positive
6  b  33.17  1.040  u  g  r  h  ...    f  0.0   t    g  00164  31285.0  positive

[5 rows x 16 columns]

# Test 1: Compare independent test
I'm going to assum that A3 is a qualification:
* Ho: the average qualification is bigger o queal for man than woman 
* H1: the average qualification is bigger for woman

In [0]:
pd.pivot_table(df,index='A1',values='A3', aggfunc={'A3':np.mean})
#A man
#B woman

A3
A1          
?   3.027917
a   5.072690
b   4.662222

In [0]:
woman['A16'] = (woman['A16'].replace({'positive':'1', 'negative':'0'})).astype(float)
man['A16'] = (man['A16'].replace({'positive':'1', 'negative':'0'})).astype(float)

In [0]:
#Ho: the average qualification is bigger o queal for man than woman man>= woman
#H1: the average qualification is bigger for woman man<woman
from scipy.stats import ttest_ind, chisquare,ttest_1samp
# compare two independent groupes column 
result=ttest_ind(man['A3'],woman['A3'],nan_policy='omit')
if result[1]<0.05:
  print(result,'\n')
  print('We reject the null hypothesis in average woman receive  qualiication is smaller than man')
else:
  print(result)
  print('My hypothesis null is correct, Man receive a bigger qualification than women\n')

Ttest_indResult(statistic=0.9892864738440149, pvalue=0.3228769597405686)
My hypothesis null is correct, Man receive a bigger qualification than women



In [0]:
woman['A3'].mean()

4.662222222222222

In [0]:
man['A3'].mean()

5.072690476190475

# TEST 2 1_Sample test

* H0: the amount  of positive and negative  are equal 

* H1: positive and negative are not equal


In [0]:
positive['A1'].mean()

0.30973451327433627

In [0]:
positive['A1'] = (positive['A1'].replace({'a':'1', 'b':'0'})).astype(float)

In [0]:
#compare 1 depend variable with mean population
#* H0: the amount  of positive and negative  are equal 
#* H1: positive and negative are not equal
A16=df['A16']
A16=(A16.replace({'positive':'1', 'negative':'0'})).astype(float)
result2= ttest_1samp(A16, 0.5, nan_policy='omit') #
print('t-statistic:',result2[0], ' p-value:', result2[1])
if result2[1] < 0.05:
  print('\n  We reject the null hypothesis.there is not the same amount of positives and negatives in the class column')
else:
  print('The null hypothesis is correct, The amount of classes positive and negative are equal')


t-statistic: -3.4370620865405166  p-value: 0.0006233978123552837

  We reject the null hypothesis.there is not the same amount of positives and negatives in the class column


#CHISQUARE

Compare 2 categorical variables and check if they are related or not 

H0: A4 is related to A16

H1: they are not related

In [0]:
contigency=pd.crosstab([df['A4']],df['A16'] )
contigency

A16  negative  positive
A4                     
l           0         2
u         263       256
y         118        45

In [0]:
result3=chi2_contingency(contigency)

In [0]:
print('stat:',result3[0])
print('p_value:',result3[1])
print('dregre Freedom:',result3[2])
print('xpected:',result3[3])
if result3[1]<0.05: 
  print('\n We reject the null hypothesis thta says that these 2 variables are related ')
else:
  print('These two variables are related')

stat: 26.234074966202144
p_value: 2.010680204180363e-06
dregre Freedom: 2
xpected: [[  1.11403509   0.88596491]
 [289.09210526 229.90789474]
 [ 90.79385965  72.20614035]]

 We reject the null hypothesis thta says that these 2 variables are related 


#CHI SQUARE Exmaple 2

I want to know if there is a a relationship between sex and a possitive or negative result in the column Class
H0: sex has a relation with the possitive or negative result 
H1: there is not relationship between these 2 variables 


In [0]:
contigency2=pd.crosstab([df['A1']],df['A16'] )
contigency2

A16  negative  positive
A1                     
a         112        98
b         262       206

In [0]:
result4=chi2_contingency(contigency2)
print('stat:',result4[0])
print('p_value:',result4[1])
print('dregre Freedom:',result4[2])
print('xpected:',result4[3])
if result4[1]<0.05: 
  print('\n We reject the null hypothesis thta says that these 2 variables are related ')
else:
  print('\n These two variables are related')

stat: 0.3112832649161995
p_value: 0.5768937883001117
dregre Freedom: 1
xpected: [[115.84070796  94.15929204]
 [258.15929204 209.84070796]]

 These two variables are related


## Part 3 - Analysis and Interpretation

Now that you've looked at the data, answer the following questions:

- Interpret and explain the two t-tests you ran - what do they tell you about the relationships between the continuous features you selected and the class labels?
- Interpret and explain the two Chi-squared tests you ran - what do they tell you about the relationships between the categorical features you selected and the class labels?
- What was the most challenging part of this sprint challenge?

Answer with text, but feel free to intersperse example code/results or refer to it from earlier.

# Test 1: Compare Ind test
I'm going to assum that A3 is a qualification:
* Ho: the average qualification is bigger o queal for man than woman 
* H1: the average qualification is bigger for woman

Ttest_indResult(statistic=0.9892864738440149, pvalue=0.3228769597405686)

My hypothesis null is correct, Man receive a bigger qualification than women because my p_value is bigger than 0.05.

I have 95% confidence that if we randomly choose a woman and a man, the qualification is going to be bigger for man than women.



# TEST 2: 1 Sample
* H0: the amount  of positive and negative  are equal 
* H1: positive and negative are not equal



t-statistic: -2.9088721445109043  
p-value: 0.0037441556293534564

We reject the null hypothesis.there is not the same amount of positives and negatives in the class column.

Assuming that my column A16 (Class) means:
positive approval
negative approval
I can assume that there is more quantity of positve/negative than the other.




#Chi-square 1:
stat: 26.234074966202144

p_value: 2.010680204180363e-06

dregre Freedom: 2

xpected: [[  1.11403509   0.88596491][289.09210526 229.90789474] [ 90.79385965  72.20614035]]


Based in my results, in which my p_value is smaller that my 0.05 confidence (95%),I can reject the idea that these 2 variables are related. 

#CHI SQUARE Example 2

I want to know if there is a a relationship between sex and a possitive or negative result in the column Class

H0: sex has a relation with the possitive or negative result 

H1: there is not relationship between these 2 variables 


stat: 0.3112832649161995

p_value: 0.5768937883001117

dregre Freedom: 1

xpected: [[115.84070796  94.15929204][258.15929204 209.84070796]]

 These two variables are related


I have a 95 % of confidence that : Sex and Class are related 
A woman is more likely to receive a negative result in the column class than man.  

**What was the most challenging part of this sprint challenge?**

This week was crazy really crazy, I used all my TL Hours every day. 
I studied really hard and when I started this sprint I had a error that I couldn't find why this was happening but I tok the Anika's advice, she told that it is much better to stop , rest for a few minutes than continue. 

And this really helped me a lot!!!


